core/src/main/java/org/apache/calcite/plan/RelOptUtil.java

```diff
@@ -3817,6 +3817,29 @@ public static RelNode pushDownJoinConditions(Join originalJoin,
         originalJoin, RelBuilder.proto(projectFactory));
   }
 
+  /**
+   * Builds a new RelNode with the pushed conditions and remaining conditions.
+   *
+   * <p> The pushed conditions will be pushed to the input of oldNode, and
+   * the remaining conditions will be keeped above the oldNode.
+   *
+   * @param builder RelBuilder
+   * @param oldNode original RelNode
+   * @param pushedConditions the conditions which pushed to the input of oldNode
+   * @param remainingConditions the conditions which keeped above the oldNode
+   */
+  public static RelNode buildRelNodeWithConditions(final RelBuilder builder, final RelNode oldNode,
+      final List<RexNode> pushedConditions, final List<RexNode> remainingConditions) {
+    RelNode rel =
+        builder.push(oldNode.getInput(0)).filter(pushedConditions).build();
+    if (rel == oldNode.getInput(0)) {
+      return oldNode;
+    }
+    rel = oldNode.copy(oldNode.getTraitSet(), ImmutableList.of(rel));
+    rel = builder.push(rel).filter(remainingConditions).build();
+    return rel;
+  }
+
```

Comment:
It's not easy to see why this is a useful utility method. Moreover, it is hard to understand what the method does without looking at the implementation.

At the moment, I would suggest reverting these changes and putting the code back inside the rule which has less exposure. Whatever we add in `RelOptUtil` class becomes public API that we need to support and maintain in every version and I am not yet sure we need this.

Reply:
Get your point,I will follow your advice.

Reply:
Now I move the `buildRelNodeWithConditions` to `FilterAggregateTransposeRule`,and in `FilterAggregateTransposeRule` and `FilterWindowTransposeRule` would use this method.

Reply:
In the final commit (https://github.com/apache/calcite/commit/bdafeec27c529d0089434a4d8b0fc7a1e49efcd1) I reverted the refactoring of `buildRelNodeWithConditions`. We can discuss in another ticket on how we can reuse code and combine the FilterAggregate and FilterWindow rules together.